identity/src/main/java/com/android/identity/MobileSecurityObjectGenerator.java

```diff
@@ -0,0 +1,312 @@
+ * as specified <em>ISO/IEC 18013-5</em> section 9.1.2 <em>Issuer data authentication</em>.
+ */
+public class MobileSecurityObjectGenerator {
+
+    private final String mDigestAlgorithm;
+    private final String mDocType;
+    private final PublicKey mDeviceKey;
+
+    private final CborBuilder mValueDigestsBuilder = new CborBuilder();
+    private final MapBuilder<CborBuilder> mValueDigestsOuter = mValueDigestsBuilder.addMap();
+    private boolean digestEmpty = true;
+
+    private final List<String> mAuthorizedNameSpaces = new ArrayList<>();
+    private final Map<String, List<String>> mAuthorizedDataElements = new HashMap<>();
+    private final Map<Integer, byte[]> mKeyInfo = new HashMap<>();
+
+    private Timestamp mSigned, mValidFrom, mValidUntil, mExpectedUpdate;
+
+    /**
+     * Constructs a new {@link MobileSecurityObjectGenerator}.
+     *
+     * @param digestAlgorithm The digest algorithm identifier. Must be one of {"SHA-256", "SHA-384", "SHA-512"}.
+     * @param docType The document type.
+     * @param deviceKey The public part of the key pair used for mdoc authentication.
+     * @exception IllegalArgumentException if the <code>digestAlgorithm</code> is not one of
+     *                                     {"SHA-256", "SHA-384", "SHA-512"}.
+     */
+    public MobileSecurityObjectGenerator(@NonNull String digestAlgorithm, @NonNull String docType,
+                                         @NonNull PublicKey deviceKey) {
+        final List<String> allowableDigestAlgorithms = List.of("SHA-256", "SHA-384", "SHA-512");
+        if (!allowableDigestAlgorithms.contains(digestAlgorithm)) {
+            throw new IllegalArgumentException("digestAlgorithm must be one of " +
+                    allowableDigestAlgorithms);
+        }
+
+        mDigestAlgorithm = digestAlgorithm;
+        mDocType = docType;
+        mDeviceKey = deviceKey;
+    }
+
+    /**
+     * Populates the <code>ValueDigests</code> mapping. This must be called at least once before
+     * generating since <code>ValueDigests</code> must be non-empty.
+     *
+     * @param nameSpace The namespace.
+     * @param digestIDs A non-empty mapping between a <code>DigestID</code> and a
+     *                  <code>Digest</code>.
+     * @return The <code>MobileSecurityObjectGenerator</code>.
+     * @exception IllegalArgumentException if the <code>digestIDs</code> is empty.
+     */
+    @NonNull
+    public MobileSecurityObjectGenerator addDigestIDs(@NonNull String nameSpace,
+            @NonNull Map<Long, byte[]> digestIDs) {
+
+        if (digestIDs.isEmpty()) {
+            throw new IllegalArgumentException("digestIDs must not be empty");
+        }
+        digestEmpty = false;
+
+        MapBuilder<MapBuilder<CborBuilder>> valueDigestsInner = mValueDigestsOuter.putMap(nameSpace);
+        for (Long digestID : digestIDs.keySet()) {
+            valueDigestsInner.put(digestID, digestIDs.get(digestID));
```

Comment:
We should verify that the `byte[]` we fish out has the right size... suggest to precompute `mDigestSize` in the constructor (32, 48, or 64) so we can use it directly here.